shared/liboxide/udev.cpp

```diff
@@ -0,0 +1,264 @@
+            O_DEBUG("UDev::Thread finished");
+        });
+        _thread.start(QThread::LowPriority);
+        moveToThread(&_thread);
+    }
+
+    UDev::~UDev(){
+        if(udevLib != nullptr){
+            udev_unref(udevLib);
+            udevLib = nullptr;
+        }
+    }
+
+    void UDev::subsystem(const QString& subsystem, std::function<void(const Device&)> callback){
+        deviceType(subsystem, NULL, callback);
+    }
+
+    void UDev::subsystem(const QString& subsystem, std::function<void()> callback){
+        deviceType(subsystem, NULL, callback);
+    }
+
+    void UDev::deviceType(const QString& subsystem, const QString& deviceType, std::function<void(const Device&)> callback){
+        connect(singleton(), &UDev::event, [callback, subsystem, deviceType](const Device& device){
+            if(device.subsystem == subsystem && (deviceType == NULL || device.deviceType == deviceType)){
+                callback(device);
+            }
+        });
+        singleton()->addMonitor(subsystem, deviceType);
+    }
+
+    void UDev::deviceType(const QString& subsystem, const QString& deviceType, std::function<void()> callback){
+        UDev::deviceType(subsystem, deviceType, [callback](const Device& device){
+            Q_UNUSED(device);
+            callback();
+        });
+    }
+
+    void UDev::start(){
+        statelock.lock();
+        O_DEBUG("UDev::Starting...");
+        exitRequested = false;
+        if(running){
+            statelock.unlock();
+            O_DEBUG("UDev::Already running");
+            return;
+        }
+        QTimer::singleShot(0, [this](){
+            monitor();
+            statelock.unlock();
+            O_DEBUG("UDev::Started");
+        });
+    }
+
+    void UDev::stop(){
+        statelock.lock();
+        O_DEBUG("UDev::Stopping...");
+        if(running){
+            exitRequested = true;
+        }
+        statelock.unlock();
+    }
+
+    bool UDev::isRunning(){ return running; }
+
+    void UDev::wait(){
+        if(isRunning()){
+            O_DEBUG("UDev::Waiting to stop...");
+            while(running){
+                qApp->processEvents();
+            }
+        }
+    }
+
+    void UDev::addMonitor(QString subsystem, QString deviceType){
+        O_DEBUG("UDev::Adding" << subsystem << deviceType);
+        QStringList* list;
+        if(monitors.contains(subsystem)){
+            list = monitors[subsystem].data();
+        }else{
+            list = new QStringList();
+            monitors.insert(subsystem, QSharedPointer<QStringList>(list));
+        }
+        if(!list->contains(deviceType)){
+            list->append(deviceType);
+            update = true;
+        }
+    }
+    void UDev::removeMonitor(QString subsystem, QString deviceType){
+        O_DEBUG("UDev::Removing" << subsystem << deviceType);
+        if(monitors.contains(subsystem)){
+            monitors[subsystem]->removeAll(deviceType);
+            update = true;
+        }
+    }
```

Comment:
Since this code isn't performance critical, I'll stick with the previous solution, as it prioritizes readability. This was quite interesting to understand though.